src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c

```diff
@@ -0,0 +1,572 @@
+        ON_EXCEPTION_PRINT_AND_GOTO(error);
+    }
+    keyPair = (*env)->NewObject(env, g_keyPairClass, g_keyPairCtor, loc[publicKey], loc[privateKey]);
+
+    goto cleanup;
+
+error:
+    if (loc[privateKey])
+    {
+        // Destroy the private key data.
+        (*env)->CallVoidMethod(env, loc[privateKey], g_destroy);
+        CheckJNIExceptions(env); // The destroy call might throw an exception. Clear the exception state.
+    }
+
+cleanup:
+    RELEASE_LOCALS(bn, CryptoNative_BigNumDestroy);
+    RELEASE_LOCALS_ENV(loc, ReleaseLRef);
+    return ToGRef(env, keyPair);
+}
+
+int32_t AndroidCryptoNative_EcKeyCreateByKeyParameters(EC_KEY** key,
+                                                const char* oid,
+                                                uint8_t* qx,
+                                                int32_t qxLength,
+                                                uint8_t* qy,
+                                                int32_t qyLength,
+                                                uint8_t* d,
+                                                int32_t dLength)
+{
+    if (!key || !oid)
+    {
+        assert(false);
+        return 0;
+    }
+
+    *key = NULL;
+
+    JNIEnv* env = GetJNIEnv();
+
+    // The easiest way to create explicit keys with a named curve is to generate
+    // new keys for the curve, pull out the explicit paramters, and then create the explicit keys.
+    *key = AndroidCryptoNative_EcKeyCreateByOid(oid);
+    if (*key == NULL)
+    {
+        return FAIL;
+    }
+
+    // Release the reference to the generated key pair. We're going to make our own with the explicit keys.
+    ReleaseGRef(env, (*key)->keyPair);
+    (*key)->keyPair =
+        AndroidCryptoNative_CreateKeyPairFromCurveParameters((*key)->curveParameters, qx, qxLength, qy, qyLength, d, dLength);
+
+    if ((*key)->keyPair == NULL)
+    {
+        // We were unable to make the keys, so clean up and return FAIL.
+        AndroidCryptoNative_EcKeyDestroy(*key);
+        *key = NULL;
+        return FAIL;
+    }
+
+    return SUCCESS;
+}
+
+// Converts a java.math.BigInteger to a positive int32_t value.
+// Returns -1 if bigInteger < 0 or > INT32_MAX
+static int32_t ConvertBigIntegerToPositiveInt32(JNIEnv* env, jobject bigInteger)
+{
+    jobject zero = (*env)->CallStaticObjectMethod(env, g_bigNumClass, g_valueOfMethod, (int64_t)0);
+    int isPositive = (*env)->CallIntMethod(env, bigInteger, g_compareToMethod, zero);
+    (*env)->DeleteLocalRef(env, zero);
+
+    // bigInteger is negative.
+    if (isPositive < 0)
+    {
+        return -1;
+    }
+
+    jobject int32MaxBigInt = (*env)->CallStaticObjectMethod(env, g_bigNumClass, g_valueOfMethod, (int64_t)INT32_MAX);
+    int willOverflow = (*env)->CallIntMethod(env, bigInteger, g_compareToMethod, int32MaxBigInt);
+    (*env)->DeleteLocalRef(env, int32MaxBigInt);
```

Comment:
In a `.java` file I'd suggest promoting this to a static field to reduce the GC cost of creating and destroying repeated copies of an immutable value.

If there's an easy way to get similar semantics from JNI without being afraid of lifetime semantics, maybe it makes sense to do so.  If you hesitate at all, it might not be worth it.

Reply:
It'll be a bit of work to get similar semantics with thread safety, so I'd rather leave it as it is at least for now.